okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/RemotingOkHttpCall.java

```diff
@@ -0,0 +1,238 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import okhttp3.Call;
+import okhttp3.Callback;
+import okhttp3.HttpUrl;
+import okhttp3.Request;
+import okhttp3.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public final class RemotingOkHttpCall extends ForwardingCall {
+
+    private static final Logger log = LoggerFactory.getLogger(RemotingOkHttpCall.class);
+
+    private static final ResponseHandler<RemoteException> remoteExceptionHandler =
+            RemoteExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<IOException> catchAllExceptionHandler = IoExceptionResponseHandler.INSTANCE;
+    private static final ResponseHandler<QosException> qosHandler = QosExceptionResponseHandler.INSTANCE;
+
+    private final BackoffStrategy backoffStrategy;
+    private final UrlSelector urls;
+    private final RemotingOkHttpClient client;
+    private final ScheduledExecutorService schedulingExecutor;
+    private final ExecutorService executionExecutor;
+    private final Duration timeout;
+
+    private final int maxNumRelocations;
+
+    RemotingOkHttpCall(
+            Call delegate,
+            BackoffStrategy backoffStrategy,
+            UrlSelector urls, RemotingOkHttpClient client,
+            ScheduledExecutorService schedulingExecutor,
+            ExecutorService executionExecutor,
+            Duration timeout, int maxNumRelocations) {
+        super(delegate);
+        this.backoffStrategy = backoffStrategy;
+        this.urls = urls;
+        this.client = client;
+        this.schedulingExecutor = schedulingExecutor;
+        this.executionExecutor = executionExecutor;
+        this.timeout = timeout;
+        this.maxNumRelocations = maxNumRelocations;
+    }
+
+    private void scheduleExecution(Callback callback, Duration backoff) {
+        schedulingExecutor.schedule(
+                () -> executionExecutor.submit(() -> doClone().enqueue(callback)),
+                backoff.toMillis(),
+                TimeUnit.MILLISECONDS);
+    }
+
+    private void doEnqueueAndHandleErrorsAndRetries(Callback callback) {
+        super.enqueue(new Callback() {
+            @Override
+            public void onFailure(Call call, IOException exception) {
+                if (!backoffStrategy.nextBackoff().isPresent()) {
+                    callback.onFailure(call, new IOException("Failed to complete the request due to an "
+                            + "IOException", exception));
+                } else {
+                    Optional<HttpUrl> redirectTo = urls.redirectToNext(request().url());
+                    if (!redirectTo.isPresent()) {
+                        callback.onFailure(call, new IOException("Failed to determine valid failover URL"
+                                + "for '" + request().url() + "' and base URLs " + urls.getBaseUrls()));
+                    } else {
```

Comment:
Could we try turning these nested if/else blocks into early returns? I think it would make it more readable and less deeply nested.

Reply:
done